EntityEncoder for ByteBuffer

cquiroz opened the pull request:
Adds an encoder for `ByteBuffer` and several tests. Fixes #870 

Somehow it feels to simple, I'm not sure if we need to do more about resource management of the buffer

rossabaker commented on core/src/main/scala/org/http4s/EntityEncoder.scala:
```diff
@@ -147,11 +147,8 @@ trait EntityEncoderInstances extends EntityEncoderInstances0 {
   implicit val byteBufferEncoder: EntityEncoder[ByteBuffer] =
-    chunkEncoder.contramap(ByteVector.view)
-   */
+    chunkEncoder.contramap(bb => Chunk.bytes(bb.array))
```
`.array` is not a guaranteed operation.  I don't do enough raw nio to know in what cases it's not supported.

scodec-bits uses a [view of a slice](https://github.com/scodec/scodec-bits/blob/0abe269dfda050cc75d27ed384e3e3a496445971/core/shared/src/main/scala/scodec/bits/ByteVector.scala#L1479-L1482).  A `Chunk[Byte]` could be invented for similar effect, but it would have very poor concatenation performance.

This might be a good design question for the fs2 experts.

cquiroz replied:
After reading a bit more you're correct, `array` is optional. It seems to me that the totally safe way is to create a `Array[Byte]` and copy the data using `get`. This is guaranteed to work and it will separate the `array` on the buffer from the one we put in the `Chunk`

It maybe terrible performance-wise though

We could check the `hasArray` method to decide to do it the other way, but changes to the original `ByteBuffer` will be reflected on the `Chunk`. scodec comment indicate that the caller should be careful about not changing the original buffer

cquiroz commented:
This turns out to be more complicated than initially thought. I added separate logic for array/non-array backed `ByteBuffer`s and added extra tests

The tests showed that some cases were not properly handled. Especially if a `ByteBuffer` is changed after the fact

Please take a look at the update and let me know what you think. I'm not totally sure it is a fully correct solution

Another question. If the `ByteBuffer` is very large, should we split it in chunks?

rossabaker commented on core/src/main/scala/org/http4s/EntityEncoder.scala:
```diff
@@ -147,11 +147,23 @@ trait EntityEncoderInstances extends EntityEncoderInstances0 {
+        Chunk.bytes(duplicate.array, 0, duplicate.remaining)
+      } else {
+        // This will create a new copy of the array, maybe expensive
+        // Should we break this if the ByteBuffer is very large?
```
I think in that case we'd have to do that as a `Process`, and lose the `contramap`.  We should still be able to provide a `Content-Length` header, unlike some of our other `Process` encoders.

I don't get deep into the NIO weeds very much.  In practice, do we see enormous non-array ByteBuffers in wild?

cquiroz replied:
I haven't seen them much but they could be useful for files. OTOH we have already a `EntityEncoder[File]`

bryce-anderson replied:
On the topic of files and `ByteBuffer`s, [`MappedByteBuffer`](https://docs.oracle.com/javase/7/docs/api/java/nio/MappedByteBuffer.html) is both interesting and offers some challenges since it will do blocking operations in our non-blocking framework. I don't have any great suggestions, just wanted to make sure everyone was aware of it.

bryce-anderson commented on core/src/main/scala/org/http4s/EntityEncoder.scala:
```diff
@@ -147,11 +147,23 @@ trait EntityEncoderInstances extends EntityEncoderInstances0 {
+        duplicate.flip()
+      }
+      if (duplicate.hasArray) {
+        Chunk.bytes(duplicate.array, 0, duplicate.remaining)
```
This isn't quite right: the buffer may have an array, but if it's also a slice or its current position isn't 0, the array offset isn't 0.

bryce-anderson replied:
Sorry, forgot to add a useful link to the docs: https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#arrayOffset()

bryce-anderson commented on core/src/main/scala/org/http4s/EntityEncoder.scala:
```diff
@@ -147,11 +147,23 @@ trait EntityEncoderInstances extends EntityEncoderInstances0 {
+      val duplicate = bb.duplicate()
+      // This is triggered if puts have been done, then we want to flip to read
+      if (duplicate.remaining != duplicate.capacity()) {
+        duplicate.flip()
```
I don't think I like this: it's not an uncommon pattern with `ByteBuffer`s for someone to read bytes from a buffer (maybe a length field or something) and then want to use the rest elsewhere.

rossabaker commented:
I think it's nice to provide a thoughtful implementation for something that's hard to get right, but if it's this hard to get universally right, I'm willing to say goodbye.  Especially because people should really be using ByteVector wherever they can.

Note that we don't try to solve the mutability problem with our array encoder.

cquiroz commented:
After the several iterations on trying to get this to work I tend to agree. It seems difficult to provide a universally correct encoder given that the underlying buffer is essentially mutable.

One way or another there will be corner cases that maybe confusing for the caller

If the encoder is not needed elsewhere inside http4s we maybe better off without it

bryce-anderson commented:
@rossabaker is right: I neglected to note that `Array[Byte]` is mutable. I think `Array[Byte]` is probably far more common, but I'd also be willing to entertain removing it as well. `ByteBuffer` is going to be far less common and has a lot more landmines such as two `ByteBuffer`s can be backed by the same underlying array, etc.

If we do keep `ByteBuffer`, I'd strongly advocate for always making a copy of the data. In addition, it's pretty common for APIs that deal in `ByteBuffer` to 'consume' the bytes you have handled, so in this case it might be proper to not use `duplicate` before making the copy so that the indexes of the buffer reflect what we have made a copy of.

If I had my druthers, I'd deprecate `EntityEncoder[ByteBuffer]` in master and not propogate it in fs2. I'm just one vote, maybe someone else has a compelling counter argument.

cquiroz commented:
Should `CharBuffer` be deprecated too?

https://github.com/http4s/http4s/blob/29d52d510bd575bf1cb1442dc19fffefdc923c3f/core/src/main/scala/org/http4s/EntityEncoder.scala#L138-L138

rossabaker commented:
Yes, I think CharBuffer would be even less commonly used and suffer all the same problems. :+1: to deprecating them both.

cquiroz commented:
Ok, I removed both
